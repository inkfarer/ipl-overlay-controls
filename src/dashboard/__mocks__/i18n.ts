import i18next from 'i18next';
import { config } from '@vue/test-utils';

jest.mock('i18next-vue', () => ({
    __esModule: true,
    useTranslation: () => ({ t: i18next.t })
}));

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
    config.global.mocks = {
        $t: i18next.t
    };
});
