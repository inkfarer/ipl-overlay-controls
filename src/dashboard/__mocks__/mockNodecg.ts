import Mock = jest.Mock;

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    listeners: {[key: string]: ReplicantChangeHandler};
    sendMessage: Mock

    constructor() {
        this.replicants = {};
        this.listeners = {};
        this.sendMessage = jest.fn();
    }

    init(): void {
        const self = this;
        window.nodecg = {
            // @ts-ignore: Just a mock
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.listeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            sendMessage: self.sendMessage,
            // @ts-ignore: Just a mock
            log: {
                info: jest.fn()
            }
        };

        // @ts-ignore: It's a mock.
        window.NodeCG = {
            waitForReplicants: jest.fn().mockResolvedValue(true)
        };
    }
}
