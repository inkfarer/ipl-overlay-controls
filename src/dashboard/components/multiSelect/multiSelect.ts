class MultiSelect extends HTMLElement {
    select: HTMLSelectElement;
    selectedElemDisplay: HTMLDivElement;
    selectedOptions: HTMLOptionElement[];

    constructor() {
        super();

        const template = document.createElement('template');
        template.innerHTML = `
            <div class="select-wrapper" id="select-wrapper">
                <slot></slot>
            </div>`;

        this.selectedOptions = [];
        this.select = this.getInnerSelect();
        this.selectedElemDisplay = this.createElemDisplay();

        // Create shadow root with styles and HTML template
        const shadow = this.attachShadow({ mode: 'open' });
        shadow.appendChild(MultiSelect.createStyle());
        shadow.appendChild(template.content.cloneNode(true));
        shadow.getElementById('select-wrapper')
            .appendChild(this.selectedElemDisplay);

        // If select element children change, check if selected elements still exist
        new MutationObserver(() => {
            const newOptions: string[] = Array.from(this.select.options)
                .map((option: HTMLOptionElement) => { return option.value; });
            this.selectedOptions = this.selectedOptions.filter(selectedOption => {
                return newOptions.includes(selectedOption.value);
            });

            this.updateElemDisplay();
            this.select.selectedIndex = -1;
        }).observe(this.select, { childList: true });
    }

    private createElemDisplay(): HTMLDivElement {
        const display = document.createElement('div');
        display.classList.add('elem-display');
        display.addEventListener('click', e => {
            // Delete selected element on click
            if ((e.target as HTMLElement).classList.contains('option')) {
                this.selectedOptions = this.selectedOptions.filter(elem => {
                    const target = e.target as HTMLDivElement;
                    return elem.innerText !== target.innerText &&
                        elem.value !== target.dataset.optionValue;
                });
                this.updateElemDisplay();
                this.dispatchEvent(new Event('change'));
            }
        });
        return display;
    }

    private getInnerSelect(): HTMLSelectElement {
        const select = this.querySelector('select');
        if (!select) {
            throw new Error('No select given to multi-select!');
        }
        select.selectedIndex = -1;

        select.addEventListener('change', event => {
            const selectedOption = select.options[select.selectedIndex];
            if (!this.selectedOptions.includes(selectedOption)) {
                this.selectedOptions.push(selectedOption);
                this.dispatchEvent(new Event('change'));
            }
            select.selectedIndex = -1;
            this.updateElemDisplay();
            event.stopPropagation();
        });

        return select;
    }

    private updateElemDisplay() {
        this.selectedElemDisplay.innerHTML = '';
        this.selectedOptions.forEach(opt => {
            const optionElem = document.createElement('div');
            optionElem.classList.add('option');
            optionElem.innerText = opt.innerText;
            optionElem.dataset.optionValue = opt.value;
            this.selectedElemDisplay.appendChild(optionElem);
        });
    }

    private static createStyle(): HTMLStyleElement {
        const style = document.createElement('style');
        /* eslint-disable max-len */
        style.textContent = `
            ::slotted(select) {
                opacity: 0;
                width: 100%;
                height: 100%;
                position: absolute;
                cursor: pointer;
                z-index: 1;
            }
            
            div.elem-display {
                border-bottom: 1px solid #737373;
                min-height: 26px;
                display: flex;
                flex-wrap: wrap;
                position: relative;
                z-index: 2;
                pointer-events: none;
                padding-bottom: 16px;
                background-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'><path fill='none' stroke='white' stroke-linecap='round' stroke-linejoin='round' stroke-width='2' d='M2 5l6 6 6-6'/></svg>");
                background-size: 16px 12px;
                background-repeat: no-repeat;
                background-position: right 3px bottom 5px;
            }
            
            div.elem-display > .option {
                background-color: #2F3A4F;
                border-radius: 8px;
                font-size: 14px;
                padding: 2px 19px 2px 4px;
                margin: 2px;
                pointer-events: auto;
                background-image: url('data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512"><!-- Font Awesome Free 5.15.3 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path fill="%23728EC2" d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"/></svg>');
                background-size: 14px 14px;
                background-repeat: no-repeat;
                background-position: right 3px center;
                transition-duration: 100ms;
            }
            
            div.elem-display > .option:hover {
                background-color: #3C4B66;
            }
            
            div.select-wrapper {
                position: relative;
                cursor: pointer;
            }`;
        /* eslint-enable max-len */

        return style;
    }
}

customElements.define('multi-select', MultiSelect);
