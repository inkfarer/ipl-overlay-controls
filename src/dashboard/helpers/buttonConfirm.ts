/*
buttonConfirm

After importing this script to a dashboard panel, add the attribute 'data-requires-confirmation' to any button.
The button will now ask for confirmation when it is first clicked.
When it is clicked a second time, a 'confirm' event is sent by it, which can be listened for by scripts.

Example:

HTML:
    <button id="dangerous-action-button" data-requires-confirmation>DO A THING</button>
JS:
    document.getElementById('dangerous-action-button').addEventListener('confirm', () => {
        // do something
    });
*/

window.addEventListener('DOMContentLoaded', () => {
    function addListener(btn: HTMLButtonElement) {
        if (btn.dataset.hasConfirmListener !== 'true') {
            btn.addEventListener('click', e => {
                const target = e.target as HTMLButtonElement;

                function resetButton() {
                    target.dataset.isClicked = 'false';
                    target.innerText = target.dataset.originalText;
                    target.style.backgroundColor = '';
                }

                if (target.dataset.isClicked === 'true') {
                    target.dispatchEvent(new Event('confirm'));
                    resetButton();
                } else {
                    target.dataset.isClicked = 'true';
                    target.dataset.originalText = target.innerText;
                    target.style.backgroundColor = 'var(--red)';
                    target.innerText = 'are you sure?';

                    setTimeout(() => {
                        resetButton();
                    }, 5000);
                }
            });

            btn.dataset.hasConfirmListener = 'true';
        }
    }

    const buttons = document.querySelectorAll('button[data-requires-confirmation]');
    buttons.forEach(addListener);

    const observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
            mutation.addedNodes.forEach(node => {
                if (node.nodeType === Node.ELEMENT_NODE
                    && node.nodeName === 'BUTTON'
                    && (node as HTMLElement).dataset.requiresConfirmation) {

                    const btn = node as HTMLButtonElement;
                    addListener(btn);
                }
            });
        });
    });

    observer.observe(document.body, {
        subtree: true,
        childList: true
    });
});
