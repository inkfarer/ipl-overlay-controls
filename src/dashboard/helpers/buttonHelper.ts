import { colors } from '../styles/colors';

export function setToggleButtonDisabled(
    toggleTrue: HTMLButtonElement,
    toggleFalse: HTMLButtonElement,
    state: boolean
): void {
    toggleTrue.disabled = state;
    toggleFalse.disabled = !state;
}

export function addChangeReminder(
    elements: HTMLElement[] | NodeListOf<HTMLElement>,
    updateButton: HTMLButtonElement
): void {
    elements.forEach(element => {
        if (!element.tagName) return;

        let event;
        if (element.tagName.toLowerCase() === 'input') event = 'input';
        else if (element.tagName.toLowerCase() === 'select') event = 'change';
        else return;

        element.addEventListener(event, e => {
            const target = e.target as HTMLElement;
            target.dataset.edited = 'true';
            updateButton.style.backgroundColor = colors.red;
        });
    });

    updateButton.addEventListener('click', () => {
        elements.forEach(elem => {
            elem.dataset.edited = 'false';
        });
        updateButton.style.backgroundColor = colors.blue;
    });
}
