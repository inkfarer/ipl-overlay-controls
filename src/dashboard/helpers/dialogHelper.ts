import { NodecgDialog } from '../types/dialog';

export function closeDialog(dialogName: string): void {
    (nodecg.getDialog(dialogName) as NodecgDialog).close();
}
