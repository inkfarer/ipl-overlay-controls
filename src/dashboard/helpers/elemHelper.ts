export function appendChildren(element: HTMLElement, ...children: HTMLElement[]): void {
    children.forEach(elem => {
        element.appendChild(elem);
    });
}

export function elementById<T extends HTMLElement>(id: string): T | undefined {
    return document.getElementById(id) as T;
}

export function dispatch(element: EventTarget, event: string): void {
    element.dispatchEvent(new Event(event));
}

export function hideElement(element: HTMLElement): void {
    element.style.display = 'none';
}

export function showElement(element: HTMLElement): void {
    element.style.removeProperty('display');
}

export function prependToBody(elem: HTMLElement): void {
    document.body.prepend(elem);
}

export function appendElementAfter(newElement: HTMLElement, element: HTMLElement): void {
    element.parentNode.insertBefore(newElement, element.nextSibling);
}
