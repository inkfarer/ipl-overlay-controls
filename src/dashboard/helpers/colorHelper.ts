import { colors, getColorOptionName } from '../../helpers/splatoonData';

export function fillColorSelector(select: HTMLSelectElement): void {
    for (let i = 0; i < colors.length; i++) {
        const element = colors[i];

        const optGroup = document.createElement('optgroup');
        optGroup.label = element.meta.name;

        for (let j = 0; j < element.colors.length; j++) {
            const color = element.colors[j];

            const option = document.createElement('option');
            option.value = getColorOptionName(color.index, element.meta.name);
            option.text = color.title;
            option.dataset.index = String(color.index);
            option.dataset.firstColor = color.clrA;
            option.dataset.secondColor = color.clrB;
            option.dataset.categoryName = element.meta.name;
            option.disabled = color.isCustom;

            optGroup.appendChild(option);
        }

        select.appendChild(optGroup);
    }
}

export function getContrastingTextColor(backgroundColor: string): string {
    const { r, g, b } = hexToRbg(backgroundColor);
    const yiq = ((r * 299) + (g * 587) + (b * 114)) / 1000;
    return (yiq >= 128) ? '#333' : 'white';
}

export function hexToRbg(color: string): { r: number, g: number, b: number } {
    color = color.replace('#', '');
    return {
        r: parseInt(color.substr(0, 2), 16),
        g: parseInt(color.substr(2, 2), 16),
        b: parseInt(color.substr(4, 2), 16)
    };
}
