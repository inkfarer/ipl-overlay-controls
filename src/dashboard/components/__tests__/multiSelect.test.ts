import '../multiSelect';
import { MultiSelect } from '../multiSelect';
import { dispatch, elementById } from '../../helpers/elemHelper';
import { UnknownFunction } from '../../../helpers/__mocks__/module';

describe('multiSelect', () => {
    it('is defined as a custom element', () => {
        expect(window.customElements.get('multi-select')).not.toBeUndefined();
    });

    it('throws error if element does not contain a select', () => {
        expect(() => new MultiSelect()).toThrow('No select given to multi-select!');
    });

    it('gets select given as element child', () => {
        document.body.innerHTML = `
        <multi-select id="ms">
            <select class="inner-select">
                <option value="opt-one"/>
                <option value="opt-two"/>
            </select>
        </multi-select>`;

        const multiSelect = elementById<MultiSelect>('ms');
        expect(multiSelect.select.options.length).toEqual(2);
        expect(multiSelect.select.classList).toContain('inner-select');
    });

    it('handles values being selected', () => {
        document.body.innerHTML = `
        <multi-select id="ms">
            <select>
                <option value="opt-one"/>
                <option value="opt-two"/>
                <option value="opt-three"/>
            </select>
        </multi-select>`;
        const multiSelect = elementById<MultiSelect>('ms');
        jest.spyOn(multiSelect, 'dispatchEvent');
        const innerSelect = multiSelect.select;
        innerSelect.value = 'opt-two';
        dispatch(innerSelect, 'change');

        expect(innerSelect.selectedIndex).toEqual(-1);
        expect(multiSelect.selectedOptions.length).toEqual(1);
        expect(multiSelect.selectedOptions[0].value).toEqual('opt-two');
        expect(multiSelect.dispatchEvent).toHaveBeenCalledWith(new Event('change'));
        expect(multiSelect.selectedElemDisplay.innerHTML).toMatchSnapshot();
    });

    it('handles the same value being selected more than once', () => {
        document.body.innerHTML = `
        <multi-select id="ms">
            <select>
                <option value="opt-one"/>
                <option value="opt-two"/>
                <option value="opt-three"/>
            </select>
        </multi-select>`;
        const multiSelect = elementById<MultiSelect>('ms');
        jest.spyOn(multiSelect, 'dispatchEvent');
        const innerSelect = multiSelect.select;
        innerSelect.value = 'opt-two';
        dispatch(innerSelect, 'change');
        innerSelect.value = 'opt-two';
        dispatch(innerSelect, 'change');

        expect(innerSelect.selectedIndex).toEqual(-1);
        expect(multiSelect.selectedOptions.length).toEqual(1);
        expect(multiSelect.selectedOptions[0].value).toEqual('opt-two');
        expect(multiSelect.dispatchEvent).toHaveBeenCalledWith(new Event('change'));
        expect(multiSelect.dispatchEvent).toHaveBeenCalledTimes(1);
        expect(multiSelect.selectedElemDisplay.innerHTML).toMatchSnapshot();
    });

    it('handles multiple values being selected', () => {
        document.body.innerHTML = `
        <multi-select id="ms">
            <select>
                <option value="opt-one"/>
                <option value="opt-two"/>
                <option value="opt-three"/>
            </select>
        </multi-select>`;
        const multiSelect = elementById<MultiSelect>('ms');
        jest.spyOn(multiSelect, 'dispatchEvent');
        const innerSelect = multiSelect.select;
        innerSelect.value = 'opt-two';
        dispatch(innerSelect, 'change');
        innerSelect.value = 'opt-three';
        dispatch(innerSelect, 'change');

        expect(innerSelect.selectedIndex).toEqual(-1);
        expect(multiSelect.selectedOptions.length).toEqual(2);
        expect(multiSelect.selectedOptions[0].value).toEqual('opt-two');
        expect(multiSelect.selectedOptions[1].value).toEqual('opt-three');
        expect(multiSelect.dispatchEvent).toHaveBeenNthCalledWith(1, new Event('change'));
        expect(multiSelect.dispatchEvent).toHaveBeenNthCalledWith(2, new Event('change'));
        expect(multiSelect.selectedElemDisplay.innerHTML).toMatchSnapshot();
    });

    it('handles deletions of selected options', () => {
        document.body.innerHTML = `
        <multi-select id="ms">
            <select>
                <option value="opt-one">owo</option>
                <option value="opt-two">owo</option>
                <option value="opt-three">owo</option>
                <option value="opt-four">uwu</option>
            </select>
        </multi-select>`;
        const multiSelect = elementById<MultiSelect>('ms');
        const innerSelect = multiSelect.select;
        innerSelect.value = 'opt-two';
        dispatch(innerSelect, 'change');
        innerSelect.value = 'opt-three';
        dispatch(innerSelect, 'change');
        innerSelect.value = 'opt-four';
        dispatch(innerSelect, 'change');

        const selectedElemDisplay = multiSelect.selectedElemDisplay;
        selectedElemDisplay.dispatchEvent(
            createClickEvent(selectedElemDisplay.querySelector('div[data-option-value="opt-two"]')));
        selectedElemDisplay.dispatchEvent(
            createClickEvent(selectedElemDisplay.querySelector('div[data-option-value="opt-four"]')));

        expect(multiSelect.selectedOptions.length).toEqual(1);
        expect(multiSelect.selectedOptions[0].value).toEqual('opt-three');
        expect(multiSelect.selectedElemDisplay.innerHTML).toMatchSnapshot();
    });

    it('handles options being changed', () => {
        let mutationObserverCallback: UnknownFunction;
        // @ts-ignore: just a test
        global.MutationObserver = class {
            constructor(callback: UnknownFunction) {
                mutationObserverCallback = callback;
            }
            observe = jest.fn();
        };

        document.body.innerHTML = `
        <multi-select id="ms">
            <select>
                <option value="opt-one">owo</option>
                <option value="opt-two">owo</option>
                <option value="opt-three">owo</option>
                <option value="opt-four">uwu</option>
            </select>
        </multi-select>`;
        const multiSelect = elementById<MultiSelect>('ms');
        const innerSelect = multiSelect.select;
        innerSelect.value = 'opt-two';
        dispatch(innerSelect, 'change');
        innerSelect.value = 'opt-three';
        dispatch(innerSelect, 'change');

        expect(multiSelect.selectedOptions.length).toEqual(2);

        innerSelect.innerHTML = `
            <option value="opt-one">owo</option>
            <option value="opt-two">owo</option>
            <option value="opt-four">uwu</option>`;

        if (!mutationObserverCallback) {
            fail('No mutation observer callback found.');
        } else {
            mutationObserverCallback();
        }

        expect(multiSelect.selectedOptions.length).toEqual(1);
    });
});

function createClickEvent(target: HTMLElement): Event {
    const result = new Event('click');
    Object.defineProperty(result, 'target', { writable: false, value: target });
    return result;
}
