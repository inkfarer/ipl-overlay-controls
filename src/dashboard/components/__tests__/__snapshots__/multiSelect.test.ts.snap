// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiSelect handles deletions of selected options 1`] = `"<div class=\\"option\\" data-option-value=\\"opt-three\\"></div>"`;

exports[`multiSelect handles multiple values being selected 1`] = `"<div class=\\"option\\" data-option-value=\\"opt-two\\"></div><div class=\\"option\\" data-option-value=\\"opt-three\\"></div>"`;

exports[`multiSelect handles the same value being selected more than once 1`] = `"<div class=\\"option\\" data-option-value=\\"opt-two\\"></div>"`;

exports[`multiSelect handles values being selected 1`] = `"<div class=\\"option\\" data-option-value=\\"opt-two\\"></div>"`;
